### PR TITLE
Validation of execution receipts

### PR DESCRIPTION
**Pull request by durkmurder**
This PR contains implementation of validation of execution receipts for consensus nodes. 
Currently it's not finished, tests are not written. Created it to get approval for selected implementation approach.

### CODE REVIEW

File: state/protocol/badger/validity.go
Comment (jordanschalm) [inline state/protocol/badger/validity.go:242]: Nice separation, very easy to read this 👏

File: state/protocol/badger/state.go
Comment (jordanschalm) [inline state/protocol/badger/state.go:17]: I'm not sure it's necessary to expose this outside of the `protocol` package. Creating a `Mutator` is an internal detail of the `protocol state` component, I don't think we would want or need to inject different implementations. IMO it would be simpler to add `results storage.ExecutionResults` to the `NewState` constructor instead and handle this factory logic internally in `Mutate()` or a helper function.

File: state/protocol/badger/validity.go
Comment (jordanschalm) [inline state/protocol/badger/validity.go:121]: I think this new receipt validation code could live in its own `receipt_validator.go` file since it's implementing a distinct component interface. The rest of this file is more helper methods for the larger protocol state.

File: state/protocol/badger/state.go
Comment (jordanschalm) [inline state/protocol/badger/state.go:28]: Do we need to instantiate a new `ReceiptValidator` each time? I believe we should be able to instantiate one instance when constructing the `State` and use that. The `protocol.State` and all `storage.*` modules should be thread-safe.

**Review by jordanschalm — COMMENTED** [review]
Nice work 👏 

The chunk verification looks good to me, but alex/martin will be better suited to check that part.

File: state/protocol/badger/validity.go
Comment (durkmurder) [reply state/protocol/badger/validity.go:121]: agree

File: state/protocol/badger/state.go
Comment (durkmurder) [reply state/protocol/badger/state.go:28]: Probably yes, but the problem is that we get access to state only in this place, and this factory gets injected into `NewState` so it should or this way or state has to be passed to factory after it's created.

File: state/protocol/badger/state.go
Comment (durkmurder) [reply state/protocol/badger/state.go:17]: We discussed this factory on the meeting, basically it's main purpose is to introduce extra data to the `Mutator` without adding extra fields to `State`. So state stays free of extra fields and new instances of `Mutator` can be injected using `MutatorFactory`

Comment (AlexHentschel) [reply state/protocol/badger/state.go:17]: @jordanschalm and I discussed this earlier today.  

This implementation here could be much simpler, if we addressed a more conceptual separation-of-concern  problem:
* When mutating a state, we need to verify the validity of the entries in the payload; currently, it is only Receipts, but there will be several more in the future. 
* For verifying an Entity, we need access to the `protocol.ReadOnlyState` 

#### Suggestion 

For Yurii's PR into `master`, I think we should **stick to the Factory pattern**, the way Yurii has currently implemented it. 

I'll attempt a refactoring of the Protocol State APIs beginning of next week (-> created [issue #5103](https://github.com/dapperlabs/flow-go/issues/5103)) and put up a draft PR.

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [reply state/protocol/badger/validity.go:121]: Generally, I was wondering about the placement of the interface 
https://github.com/onflow/flow-go/blob/d9d7f00ff2e9304f6c9e19088161ba40a1575123/state/protocol/validation.go#L5-L7
as well as its concrete implementation 
https://github.com/onflow/flow-go/blob/7c3ceeddf342532f0a073f5e2ee2cad5cbb947ce/state/protocol/badger/validity.go#L121-L126

Here are my thoughts:
* The verification components, like `ReceiptValidator` are not specific to the protocol state. There should (or will be) other component which need to verify the validity of Execution Receipts. 
* we have other general service components, such as a [verification component for signatures](https://github.com/onflow/flow-go/blob/7c3ceeddf342532f0a073f5e2ee2cad5cbb947ce/module/verifier.go#L8) in our [`module`](https://github.com/onflow/flow-go/tree/7c3ceeddf342532f0a073f5e2ee2cad5cbb947ce/module) package. Accordingly, we could place the entire content of 
   - interface `ReceiptValidator` into `module/receipt_validator.go`
   - implementation `receiptValidator` into `module/validation/receipt_validator.go`

File: state/protocol/badger/state.go
Comment (AlexHentschel) [reply state/protocol/badger/state.go:28]: It feels a little awkward, I admit. Nevertheless, I think this is the best starting point for me to work towards [issue #5103](https://github.com/dapperlabs/flow-go/issues/5103)

**Review by AlexHentschel — COMMENTED** [review]
Some high-level comments regarding code structure.

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:133]: Echoing same sentiment as in my [earlier comment](https://github.com/onflow/flow-go/pull/180#discussion_r530731295):
* for testability the implementation of `module.Verifier` should be injected into the `receiptValidator`

Regarding the specific implementation type that you need to use: 
* Check the data structure, whose signature you want to verify: https://github.com/onflow/flow-go/blob/d9d7f00ff2e9304f6c9e19088161ba40a1575123/model/flow/execution_receipt.go#L28-L33

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:180]: ⚠️ 
* signing the ID of the receipt does not work:
   - the receipt itself contains the creator's signature, which would create a circular dependency 
   - instead, the Execution Node would sign the ID of the `ExecutionReceipt.Body()`
* the execution node currently does not sign its receipts 😕 
https://github.com/onflow/flow-go/blob/d9d7f00ff2e9304f6c9e19088161ba40a1575123/engine/execution/ingestion/engine.go#L1276 


Generally, I would suggest:
*  we make a separate issue for the cryptography; implementing it simultaneously in Execution and Consensus node
    -  Create dedicated structs for signing and verifying Execution Receipts in a central place, which are used by Execution node and other nodes; the implementation could follow [`SingleSigner ` and  `SingleSignerVerifier` for consensus messages](https://github.com/onflow/flow-go/blob/d9d7f00ff2e9304f6c9e19088161ba40a1575123/consensus/hotstuff/verification/single_signer.go)

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:195]: should be `InvalidInputError`

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:199]: should be `InvalidInputError`

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:221]: should be `InvalidInputError`

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:231]: I think we need to explicitly handle the sentinel error: `storage.ErrNotFound` if there is no result such known. For now, I think we can treat the receipt as invalid (we might want to further differentiate in the future), i.e. we can return a `InvalidInputError`

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:234]:
```suggestion

```

File: state/protocol/badger/validity.go
Comment (AlexHentschel) [inline state/protocol/badger/validity.go:237]: to reduce dependency between the different methods (rather, in which order they have to be executed), I would suggest we also handle the possibility of `storage.ErrNotFound` error.
Thereby, there is no implicit assumption about in which order we call the verification methods `verifySignature`, `verifyChunksFormat`, `verifyExecutionResult`

**Review by AlexHentschel — COMMENTED** [review]
some more detailed comments regarding the implementation details in `receiptValidator`

File: engine/consensus/matching/engine.go
Comment (durkmurder) [reply engine/consensus/matching/engine.go:106]: Agree, currently working on tests.

File: state/protocol/badger/validity.go
Comment (durkmurder) [reply state/protocol/badger/validity.go:121]: I like it, will do that way if everyone is ok with it.

File: state/protocol/badger/validity.go
Comment (durkmurder) [reply state/protocol/badger/validity.go:170]: yes, we discussed this, I believe at this point we treat everything as error and will do that later. I can add TODO for this one.

File: state/protocol/badger/validity.go
Comment (durkmurder) [reply state/protocol/badger/validity.go:186]: I was checking [here](https://github.com/onflow/flow-go/blob/master/engine/execution/ingestion/engine.go#L1287) seems like we have a signature, but indeed I don't like how it's built at the moment.

File: state/protocol/badger/mutator.go
Comment (durkmurder) [reply state/protocol/badger/mutator.go:607]: I would like to avoid using `(bool, error)` because I agree that it's not informative. I am a fan of returning error if something goes wrong, using sentinel values or `nil` if everything went good. I like your proposal and will do that way.

File: module/validation/receipt_validator.go
Comment (AlexHentschel) [inline module/validation/receipt_validator.go:138]: for `Identifier`, we generally use the verb `%v` (or `%x`) ... rarely `%s` (though, I think it should work). @jordanschalm would know best

File: module/validation/receipt_validator.go
Comment (AlexHentschel) [inline module/validation/receipt_validator.go:129]:
```suggestion
			return engine.NewInvalidInputErrorf("receipt's previous result (%x) is unknown")
```

File: module/validation/receipt_validator.go
Comment (AlexHentschel) [inline module/validation/receipt_validator.go:15]: I think we generally export implementations of interfaces, and return their specific type at the constructor:
```golang
func NewReceiptValidator(state protocol.ReadOnlyState, index storage.Index, results storage.ExecutionResults) receiptValidator
``` 
(just to have a consistent style in the code base)

**Review by AlexHentschel — APPROVED** [review]
Looks good to me. 
* Some minor comments regarding style 
* Lets make sure we get `NewInvalidExtensionErrorf` correct (see comment below)

File: state/protocol/badger/mutator.go
Comment (AlexHentschel) [inline state/protocol/badger/mutator.go:245]: not sure this is needed. By wrapping
https://github.com/onflow/flow-go/blob/c5671a21f466bcef93fd105277ee827f490f581c/state/protocol/badger/mutator.go#L246
the error type is preserved.

For example:
```golang
        err0 := NewInvalidExtensionError(msg string)
        err1 := fmt.Errorf("payload receipts not compliant with chain state: %w", err0)
 
        state.IsInvalidExtensionError(err0) // true
        state.IsInvalidExtensionError(err1) // true
```

In fact, it is good practise to wrap errors to make expose information about the error trace. In other words, even if `err` is an `InvalidExtensionError`, it would still be preferred to wrap the error here:
```golang
return fmt.Errorf("payload receipts not compliant with chain state: %w", err0)
```

I am sure you have reasons for wrapping the error / not wrapping the error depending on its type. Maybe my [earlier suggestion](https://github.com/onflow/flow-go/pull/180#discussion_r531259144) would also remove the need to not wrap the error here?

File: state/protocol/badger/mutator.go
Comment (AlexHentschel) [inline state/protocol/badger/mutator.go:620]:
```suggestion
				return state.NewInvalidExtensionErrorf("some entities referenced by receipt %v are missing: %w", receipt.ID(), err)
```

File: module/validation/receipt_validator.go
Comment (jordanschalm) [reply module/validation/receipt_validator.go:138]: They all work :) 
https://github.com/onflow/flow-go/blob/1e0a30e7a9118aad96a706639e37ec955f562542/model/flow/identifier.go#L44-L53

File: module/validation/receipt_validator.go
Comment (zhangchiqing) [inline module/validation/receipt_validator.go:164]: A good practice [is to wrap the error with more context](https://github.com/onflow/flow-go/blob/1efef2e885095890bcc9682b9638591fedcd30ba/state/protocol/badger/mutator.go#L93-L110), so that the error message could include the stack trace.

File: state/errors.go
Comment (AlexHentschel) [reply state/errors.go:10]: Inspired by a pattern Max Walter introduced:
https://github.com/onflow/flow-go/blob/c5671a21f466bcef93fd105277ee827f490f581c/engine/errors.go#L10-L42

I suggested this approach:
* it allows to wrap lower-level errors as `InvalidExtensionError` and have the higher-level business logic only check for `InvalidExtensionError` 
* I find this fairly convenient, as it allows us to retain the lower-level error trace (with all the auxiliary information), which would not be the case, if we just retained the message.

**Comment by durkmurder** [comment]
@AlexHentschel @jordanschalm @zhangchiqing I've added tests and made changes that were mentioned in comments. Waiting for final review.

**Review by arrivets — APPROVED** [review]

**Pull request merged**
